packages/financial-templates-lib/src/price-feed/UniswapPriceFeed.js

```diff
@@ -227,15 +244,47 @@ class UniswapPriceFeed extends PriceFeedInterface {
       return reserve1.mul(fixedPointAdjustment).div(reserve0);
     }
   }
+}
 
-  _computeTwap(eventsIn, startTime, endTime) {
-    const events = eventsIn.map(e => {
-      return [e.timestamp, e.price];
+class UniswapV3PriceFeed extends UniswapPriceFeed {
+  async _getSortedEvents(fromBlock, toBlock) {
+    const events = await this.uniswap.getPastEvents("Swap", { fromBlock: fromBlock, toBlock: toBlock });
+    // Primary sort on block number. Secondary sort on transactionIndex. Tertiary sort on logIndex.
+    events.sort((a, b) => {
+      if (a.blockNumber !== b.blockNumber) {
+        return a.blockNumber - b.blockNumber;
+      }
+
+      if (a.transactionIndex !== b.transactionIndex) {
+        return a.transactionIndex - b.transactionIndex;
+      }
+
+      return a.logIndex - b.logIndex;
     });
-    return computeTWAP(events, startTime, endTime, this.toBN("0"));
+
+    return events;
+  }
+
+  _getPriceFromEvent(event) {
+    // Fixed point adjustment should use same precision as token0, unless price is inverted.
+    const fixedPointAdjustment = this.toBN(
+      parseFixed("1", this.invertPrice ? this.token1Precision : this.token0Precision).toString()
+    );
+
+    const X96Adjustment = this.toBN(2).pow(this.toBN(96));
+    const rawSqrtPrice = this.toBN(event.returnValues.sqrtPriceX96);
+    // This effectively computes the price by squaring the value, multiplying it up by our intended precision, then dividing out both X96 fixed point multipliers.
+    const nonInvertedPrice = rawSqrtPrice
+      .mul(rawSqrtPrice)
+      .mul(fixedPointAdjustment)
+      .div(X96Adjustment)
```

Comment:
this double multiplication and division seems overly complex. why don't you just use `pow`? for power? In my experimentation on the range trader, I simply had a method like this
```
function decodePriceSqrt(priseSqrt) {
  return new bn(priseSqrt).dividedBy(new bn(2).pow(96)).pow(2);
}
```

Reply:
I don't use bignumber.js (handles decimals) to avoid having to convert back and forth and to avoid adding an extra runtime dependency.

Without using it, we have to multiply up before dividing to preserve precision. However, we could switch the double multiplies with using pow, but I'm not sure how much simpler it would be. Here's a sample of that implementation:
```
rawSqrtPrice.pow(this.toBN(2)).mul(fixedPointAdjustment).div(this.toBN(2).pow(192));
```

Personally, I don't find this to be much simpler or any more concise (even if you were to sub in X96Adjustment instead of doing the pow). I find it a bit more difficult to decode what's going on. However, if you find something like this to be easier to read and interpret I'm happy to change, as I may be in the minority since I may be biased!

Reply:
ye, it's a good point. having to pull in big number.js is not ideal. I'm ok with leaving it how you've got it here to avoid that dependency within the price feed.

In the unit tests I've pulled it in but there is also a fair amount of this kind of computation there so I feel ok with having it in that context.

Reply:
Yep, and same, I pulled it into dev deps for unit tests, but wanted to avoid having it installed with the package if it could be avoided.